File: backend/app/schemas/admin_analytics.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class AnalyticsPeriod(BaseModel):
    days: int
    start: datetime
    end: datetime


class AnalyticsSummary(BaseModel):
    revenue: Decimal
    orders: int
    customers: int
    products_sold: int
    average_order_value: Decimal
    revenue_change: float
    orders_change: float
    customers_change: float
    products_sold_change: float


class SalesTrendItem(BaseModel):
    date: datetime
    revenue: Decimal
    orders: int


class TopProductAnalytics(BaseModel):
    product_id: str
    product_name: str
    quantity: int
    revenue: Decimal


class CategoryAnalytics(BaseModel):
    category_id: str
    category_name: str
    revenue: Decimal
    quantity: int
    percentage: Decimal


class CustomerAnalytics(BaseModel):
    new_customers: int
    returning_customers: int
    total_customers: int
    returning_rate: Decimal


class AdminAnalyticsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    period: AnalyticsPeriod
    summary: AnalyticsSummary
    sales_trend: list[SalesTrendItem]
    top_products: list[TopProductAnalytics]
    categories: list[CategoryAnalytics]
    customers: CustomerAnalytics

File: frontend/lib/api/admin-analytics.ts
import api from "./client";

export type AnalyticsPeriod = {
  days: number;
  start: string;
  end: string;
};

export type AnalyticsSummary = {
  revenue: string | number;
  orders: number;
  customers: number;
  products_sold: number;
  average_order_value: string | number;
  revenue_change: number;
  orders_change: number;
  customers_change: number;
  products_sold_change: number;
};

export type SalesTrendItem = {
  date: string;
  revenue: string | number;
  orders: number;
};

export type TopProductAnalytics = {
  product_id: string;
  product_name: string;
  quantity: number;
  revenue: string | number;
};

export type CategoryAnalytics = {
  category_id: string;
  category_name: string;
  revenue: string | number;
  quantity: number;
  percentage: string | number;
};

export type CustomerAnalytics = {
  new_customers: number;
  returning_customers: number;
  total_customers: number;
  returning_rate: string | number;
};

export type AdminAnalytics = {
  period: AnalyticsPeriod;
  summary: AnalyticsSummary;
  sales_trend: SalesTrendItem[];
  top_products: TopProductAnalytics[];
  categories: CategoryAnalytics[];
  customers: CustomerAnalytics;
};

export const adminAnalyticsApi = {
  get: (days = 30) =>
    api.get<AdminAnalytics>("/admin/analytics", {
      params: { days },
    }),
};
